data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/LogicalReplicationEventProcessor.java

```diff
@@ -0,0 +1,262 @@
+        char messageType = (char) msg.get();
+        if (messageType == 'B') {
+            processBeginMessage(msg);
+        } else if (messageType == 'R') {
+            processRelationMessage(msg);
+        } else if (messageType == 'I') {
+            processInsertMessage(msg);
+        } else if (messageType == 'U') {
+            processUpdateMessage(msg);
+        } else if (messageType == 'D') {
+            processDeleteMessage(msg);
+        } else if (messageType == 'C') {
+            processCommitMessage(msg);
+        } else {
+            throw new IllegalArgumentException("Replication message type [" + messageType + "] is not supported. ");
+        }
+    }
+
+    void processBeginMessage(ByteBuffer msg) {
+        currentLsn = msg.getLong();
+        long epochMicro = msg.getLong();
+        currentEventTimestamp = convertPostgresEventTimestamp(epochMicro);
+        int transaction_xid = msg.getInt();
+
+        LOG.debug("Processed BEGIN message with LSN: {}, Timestamp: {}, TransactionId: {}", currentLsn, currentEventTimestamp, transaction_xid);
+    }
+
+    void processRelationMessage(ByteBuffer msg) {
+        int tableId = msg.getInt();
+        // null terminated string
+        String schemaName = getNullTerminatedString(msg);
+        String tableName = getNullTerminatedString(msg);
+        int replicaId = msg.get();
+        short numberOfColumns = msg.getShort();
+
+        List<String> columnNames = new ArrayList<>();
+        for (int i = 0; i < numberOfColumns; i++) {
+            int flag = msg.get();    // 1 indicates this column is part of the replica identity
+            // null terminated string
+            String columnName = getNullTerminatedString(msg);
+            ColumnType columnType = ColumnType.getByTypeId(msg.getInt());
+            String columnTypeName = columnType.getTypeName();
+            int typeModifier = msg.getInt();
+            if (columnType == ColumnType.VARCHAR) {
+                int varcharLength = typeModifier - 4;
+            } else if (columnType == ColumnType.NUMERIC) {
+                int precision = (typeModifier - 4) >> 16;
+                int scale = (typeModifier - 4) & 0xFFFF;
+            }
+            columnNames.add(columnName);
+        }
+
+        // TODO: get primary keys in advance
+        final TableMetadata tableMetadata = new TableMetadata(
+                tableName, schemaName, columnNames, List.of("id"));
+
+        tableMetadataMap.put((long) tableId, tableMetadata);
+
+        LOG.debug("Processed an Relation message with RelationId: {} Namespace: {} RelationName: {} ReplicaId: {}", tableId, schemaName, tableName, replicaId);
+    }
+
+    void processCommitMessage(ByteBuffer msg) {
+        int flag = msg.get();
+        long commitLsn = msg.getLong();
+        long endLsn = msg.getLong();
+        long epochMicro = msg.getLong();
+
+        if (currentLsn != commitLsn) {
+            // This shouldn't happen
```

Comment:
Should we consider throwing an exception instead of just logging ?

Reply:
Updated to throw an exception.